File: packages/source-indexer/indexer/relations.ts
import type { ProjectDefinition, ProjectRelation } from '@crux/core/catalog'
import { projectRelation } from './relation-registry'
import type { StaticFoundDefinition } from './types'

export function relationsFromStaticDefinitions(
  found: readonly StaticFoundDefinition[],
  importedDefinitions = new Map<string, ProjectDefinition>(),
): ProjectRelation[] {
  const byVariable = new Map(found.map((item) => [item.variableName, item.definition]))
  return found.flatMap((item) =>
    item.relationRefs.flatMap((ref) => {
      const source = ref.fromVariable ? (byVariable.get(ref.fromVariable) ?? importedDefinitions.get(ref.fromVariable)) : undefined
      const target = ref.toVariable ? (byVariable.get(ref.toVariable) ?? importedDefinitions.get(ref.toVariable)) : undefined
      const targetId = ref.toId ?? target?.id ?? fallbackRelationTargetId(ref.type, ref.toVariable)
      const sourceId = ref.fromId ?? source?.id ?? item.definition.id
      const type = target?.kind && ref.typeByTargetKind?.[target.kind] ? ref.typeByTargetKind[target.kind] : ref.type
      if (!targetId || !type) return []
      const sourceFidelity = ref.fromId ? item.definition.fidelity : (source?.fidelity ?? item.definition.fidelity)
      const targetFidelity = ref.toId ? 'resolved' : target?.fidelity
      const fidelity = sourceFidelity === 'resolved' && targetFidelity === 'resolved' ? 'resolved' : 'partial'
      return [
        projectRelation({
          type,
          from: sourceId,
          to: targetId,
          fidelity,
          source: item.definition.source,
        }),
      ]
    }),
  )
}

function fallbackRelationTargetId(type: string, variableName: string | undefined): string | undefined {
  if (!variableName) return undefined
  switch (type) {
    case 'agent.uses_prompt':
    case 'flow.step.uses_prompt':
      return `prompt:${safeVariableId(variableName)}`
    case 'agent.uses_tool':
    case 'flow.step.uses_tool':
      return `tool:${variableName}`
    case 'agent.reads_memory':
    case 'agent.writes_memory':
    case 'prompt.reads_memory':
    case 'prompt.writes_memory':
    case 'context.reads_memory':
    case 'context.writes_memory':
    case 'tool.reads_memory':
    case 'tool.writes_memory':
      return `memory:${safeVariableId(variableName)}`
    case 'agent.reads_blackboard':
    case 'agent.writes_blackboard':
    case 'prompt.reads_blackboard':
    case 'prompt.writes_blackboard':
    case 'context.reads_blackboard':
    case 'context.writes_blackboard':
    case 'tool.reads_blackboard':
    case 'tool.writes_blackboard':
      return `blackboard:${safeVariableId(variableName)}`
    case 'agent.reads_workspace':
    case 'agent.writes_workspace':
    case 'prompt.reads_workspace':
    case 'prompt.writes_workspace':
    case 'context.reads_workspace':
    case 'context.writes_workspace':
    case 'tool.reads_workspace':
    case 'tool.writes_workspace':
      return `workspace:${safeVariableId(variableName)}`
    case 'flow.step.uses_agent':
      return `agent:${safeVariableId(variableName)}`
    case 'flow.step.uses_memory':
    case 'flow.step.reads_memory':
    case 'flow.step.writes_memory':
    case 'swarm.uses_memory':
      return `memory:${safeVariableId(variableName)}`
    case 'flow.step.uses_blackboard':
    case 'flow.step.reads_blackboard':
    case 'flow.step.writes_blackboard':
    case 'swarm.uses_blackboard':
      return `blackboard:${safeVariableId(variableName)}`
    case 'flow.step.reads_workspace':
    case 'flow.step.writes_workspace':
      return `workspace:${safeVariableId(variableName)}`
    case 'composition.uses_prompt':
    case 'parallel.branch.uses_prompt':
    case 'pipeline.stage.uses_prompt':
      return `prompt:${safeVariableId(variableName)}`
    case 'composition.uses_tool':
    case 'parallel.branch.uses_tool':
    case 'pipeline.stage.uses_tool':
    case 'workspace.exposes_tool':
      return `tool:${variableName}`
    case 'composition.uses_flow':
    case 'parallel.branch.uses_flow':
    case 'pipeline.stage.uses_flow':
      return `flow:${safeVariableId(variableName)}`
    case 'composition.uses_agent':
    case 'parallel.branch.uses_agent':
    case 'pipeline.stage.uses_agent':
    case 'consensus.includes_agent':
    case 'swarm.includes_agent':
      return `agent:${safeVariableId(variableName)}`
    case 'agent.uses_routing':
    case 'flow.step.uses_routing':
    case 'composition.uses_routing':
    case 'parallel.branch.uses_routing':
    case 'pipeline.stage.uses_routing':
      return `routing.router:${safeVariableId(variableName)}`
    case 'consensus.uses_scorer':
    case 'rag.pipeline.stage.uses_scorer':
      return `scorer:${safeVariableId(variableName)}`
    case 'consensus.uses_judge':
      return `agent:${safeVariableId(variableName)}`
    case 'rag.pipeline.uses_retriever':
    case 'rag.pipeline.stage.uses_retriever':
      return `rag.retriever:${safeVariableId(variableName)}`
    case 'constraint.applies_to':
    case 'guardrail.applies_to':
    case 'eval.covers_definition':
      return variableName.includes(':') ? variableName : undefined
    default:
      return undefined
  }
}

function safeVariableId(value: string): string {
  return value
    .replace(/([a-z0-9])([A-Z])/g, '$1-$2')
    .replace(/([A-Z])([A-Z][a-z])/g, '$1-$2')
    .toLowerCase()
    .replace(/[^a-z0-9_.:-]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

File: crates/primitives/src/composition/relations.rs
use serde_json::{Value, json};

use crate::{
    composition::output::{CompositionChild, composition_child_relation_refs},
    composition::values::{identifier_array, pipeline_stage_target},
    context::PrimitiveContext,
    definition::safe_id,
    protocol::StaticSyntaxValue,
    record_values::{
        direct_string_property, object_array_value, object_map_identifier_entries, property_value,
        reference_property,
    },
};

pub(crate) fn composition_references(
    context: &PrimitiveContext<'_>,
    call_name: &str,
    config: &StaticSyntaxValue,
    composition_id: &str,
    children: &[CompositionChild],
) -> Vec<Value> {
    let mut refs = composition_agent_refs(context, call_name, config)
        .into_iter()
        .map(composition_uses_ref)
        .collect::<Vec<_>>();
    refs.extend(structural_relation_refs(
        context,
        call_name,
        config,
        composition_id,
    ));
    refs.extend(composition_child_relation_refs(call_name, children));
    refs
}

fn composition_agent_refs(
    context: &PrimitiveContext<'_>,
    call_name: &str,
    config: &StaticSyntaxValue,
) -> Vec<String> {
    match call_name {
        "parallel" | "swarm" => {
            object_map_identifier_entries(property_value(config, "agents"), &context.initializers)
                .into_iter()
                .map(|(_, value)| value)
                .collect()
        }
        "consensus" => identifier_array(context, config, "agents"),
        "pipeline" => object_array_value(property_value(config, "steps"), &context.initializers)
            .into_iter()
            .filter_map(|stage| pipeline_stage_target(context, stage).map(|target| target.variable))
            .collect(),
        _ => Vec::new(),
    }
}

fn structural_relation_refs(
    context: &PrimitiveContext<'_>,
    call_name: &str,
    config: &StaticSyntaxValue,
    composition_id: &str,
) -> Vec<Value> {
    match call_name {
        "consensus" => consensus_relation_refs(context, config, composition_id),
        "swarm" => swarm_relation_refs(context, config, composition_id),
        _ => Vec::new(),
    }
}

fn consensus_relation_refs(
    context: &PrimitiveContext<'_>,
    config: &StaticSyntaxValue,
    composition_id: &str,
) -> Vec<Value> {
    let mut refs = identifier_array(context, config, "agents")
        .into_iter()
        .map(|to_variable| json!({"type": "consensus.includes_agent", "fromId": composition_id, "toVariable": to_variable}))
        .collect::<Vec<_>>();
    if let Some(judge) = reference_property(config, "judge", &context.initializers) {
        refs.push(json!({
            "type": "consensus.uses_judge",
            "typeByTargetKind": { "agent": "consensus.uses_judge", "scorer": "consensus.uses_scorer" },
            "fromId": composition_id,
            "toVariable": judge,
        }));
    }
    if let Some(scorer) = reference_property(config, "scorer", &context.initializers) {
        refs.push(json!({"type": "consensus.uses_scorer", "fromId": composition_id, "toVariable": scorer}));
    }
    refs
}

fn swarm_relation_refs(
    context: &PrimitiveContext<'_>,
    config: &StaticSyntaxValue,
    composition_id: &str,
) -> Vec<Value> {
    let mut refs = object_map_identifier_entries(property_value(config, "agents"), &context.initializers)
        .into_iter()
        .map(|(_, value)| json!({"type": "swarm.includes_agent", "fromId": composition_id, "toVariable": value}))
        .collect::<Vec<_>>();
    if let Some(coordinator) = direct_string_property(config, "startAgent") {
        refs.push(json!({"type": "swarm.coordinated_by", "fromId": composition_id, "toId": format!("agent:{}", safe_id(&coordinator))}));
    }
    if let Some(blackboard) = reference_property(config, "blackboard", &context.initializers) {
        refs.push(json!({"type": "swarm.uses_blackboard", "fromId": composition_id, "toVariable": blackboard}));
    }
    if let Some(memory) = reference_property(config, "memory", &context.initializers) {
        refs.push(
            json!({"type": "swarm.uses_memory", "fromId": composition_id, "toVariable": memory}),
        );
    }
    refs.extend(identifier_array(context, config, "memory").into_iter().map(
        |memory| json!({"type": "swarm.uses_memory", "fromId": composition_id, "toVariable": memory}),
    ));
    refs
}

fn composition_uses_ref(to_variable: String) -> Value {
    json!({
        "type": "composition.uses_agent",
        "typeByTargetKind": {
            "agent": "composition.uses_agent",
            "flow": "composition.uses_flow",
            "prompt": "composition.uses_prompt",
            "tool": "composition.uses_tool",
            "routing.router": "composition.uses_routing",
            "routing.cascade": "composition.uses_routing",
            "routing.fallback": "composition.uses_routing",
        },
        "toVariable": to_variable,
    })
}
